### PR TITLE
Bindings: Allow coercion between Vector3 subtypes.

### PR DESCRIPTION
**Pull request by madmaxoft**
In manually bound functions, allows one to use any `Vector3<T>` value, as well as a `{x, y, z}` table, in Lua for any `Vector3<T>` parameter.
Has example in Debuggers' `/vector` command.
Unfortunately doesn't work in auto-bindings.

### CODE REVIEW

File: src/Bindings/ManualBindings_BlockArea.cpp [hunk 1]
```diff
@@ -48,14 +48,12 @@ static int readCuboidOverloadParams(cLuaState & a_LuaState, int a_StartParam, cC
 	else
 	{
 		// Assume the 2-Vector3i version:
-		Vector3i * p1;
-		Vector3i * p2;
-		if (!a_LuaState.GetStackValues(a_StartParam, p1, p2))
+		Vector3i p1;
+		Vector3i p2;
```
Comment (bearbin) [inline src/Bindings/ManualBindings_BlockArea.cpp:52]: These are no longer used?

Comment (madmaxoft) [reply src/Bindings/ManualBindings_BlockArea.cpp:52]: No, just temporaries

**Comment by madmaxoft** [comment]
Unfortunately this doesn't want to build with Clang due to the template instantiation deferred to the cpp file; I'm not friends enough with templates to fix this. Perhaps @worktycho or @peterbell10 with their extensive knowledge could step in?

File: src/Bindings/ManualBindings_BlockArea.cpp (hunk 1 above)
Comment (peterbell10) [reply src/Bindings/ManualBindings_BlockArea.cpp:52]: You use `a_Cuboid.p1` directly now, these variables are never used.

**Comment by peterbell10** [comment]
It's only a warning, not a build error. Could just add a `-Wno-error=undefined-func-template` to the `CMakeLists.txt`.

Alternatively, you could have the `Vector3<lua_Number>` version in the cpp file and base the template off of that. With the full template definition in the header file.

**Comment by madmaxoft** [comment]
Declaring the explicit instantiation as `extern` in the header seems to work as well.

**Comment by madmaxoft** [comment]
Ready for review.
Now I only wish the same thing could somehow be done to the auto-bindings...

File: src/Bindings/BindingsProcessor.lua [hunk 2]
```diff
@@ -231,38 +246,40 @@ local function OutputLuaStateHelpers(a_Package)
...
-			f:write("\t\treturn false;\n\t}\n")
-			f:write("\ttolua_Error err;\n")
-			f:write("\tif (tolua_isusertype(m_LuaState, a_StackPos, \"const " .. item.name .. "\", false, &err))\n")
-			f:write("\t{\n")
-			f:write("\t\ta_ReturnedVal = *(static_cast<const " .. item.name .. " **>(lua_touserdata(m_LuaState, a_StackPos)));\n")
-			f:write("\t\treturn true;\n");
-			f:write("\t}\n")
-			f:write("\treturn false;\n")
-			f:write("}\n\n\n\n\n\n")
+			if not(g_HasCustomGetImplementation[item.lname]) then
```
Comment (tigerw) [inline src/Bindings/BindingsProcessor.lua:256]: Does Lua support inverting conditions?

File: src/Bindings/DeprecatedBindings.cpp [hunk 3]
```diff
@@ -604,21 +561,6 @@ void DeprecatedBindings::Bind(lua_State * tolua_S)
...
-	tolua_beginmodule(tolua_S, "Vector3f");
-		tolua_function(tolua_S,"abs",   tolua_Vector3_Abs<float>);
-		tolua_function(tolua_S,"clamp", tolua_Vector3_Clamp<float>);
-	tolua_endmodule(tolua_S);
-
-	tolua_beginmodule(tolua_S, "Vector3d");
-		tolua_function(tolua_S,"abs",   tolua_Vector3_Abs<double>);
-		tolua_function(tolua_S,"clamp", tolua_Vector3_Clamp<double>);
-	tolua_endmodule(tolua_S);
-
```
Comment (tigerw) [inline src/Bindings/DeprecatedBindings.cpp:621]: yes, YES!

**Comment by tigerw** [comment]
So am I correct in saying the new specialisations for GetStackValue are what enable both the coercion, and the treatment of Vector3 as not just an opaque pointer in the manual bindings?

File: src/Bindings/BindingsProcessor.lua (hunk 2 above)
Comment (madmaxoft) [reply src/Bindings/BindingsProcessor.lua:256]: What do you mean? Invert the condition and `continue`? Lua doesn't have a `continue`, unfortunately.

**Comment by madmaxoft** [comment]
I don't quite understand what you mean, but the changes here allow you to use any kind of `Vector3` as a parameter to any function accepting any `Vector3`, as long as the function is bound manually. The example in Debuggers shows this. Nothing more, nothing less.

File: src/Bindings/BindingsProcessor.lua (hunk 2 above)
Comment (tigerw) [reply src/Bindings/BindingsProcessor.lua:256]: aw shucks

**Comment by tigerw** [comment]
Just noting the Vector3 pointer + null check ↝ plain Vector3 changes were great for readability.

File: src/Bindings/ManualBindings.cpp [hunk 4]
```diff
@@ -2975,8 +2975,8 @@ static int tolua_cLineBlockTracer_FirstSolidHitTrace(lua_State * tolua_S)
 		}
 		// Get the params:
 		cWorld * world;
-		Vector3d * start;
-		Vector3d * end;
+		Vector3d start;
+		Vector3d end;
 		if (!L.GetStackValues(idx, world, start, end))
```
Comment (peterbell10) [inline src/Bindings/ManualBindings.cpp:2980]: Above this is `L.IsParamUserType(idx + 1, "Vector3<double>")` and `L.CheckParamUserType(idx + 1, "Vector3<double>", idx + 2)`. I think you need a `CheckParamVector3`.

Comment (madmaxoft) [reply src/Bindings/ManualBindings.cpp:2980]: What an unbelievable catch

Comment (peterbell10) [reply src/Bindings/ManualBindings.cpp:2980]: You didn't address this fully. You didn't add a `CheckParamVector3` and if you search for `Vector3<double>` then you can see there is also an issue with `tolua_cLineBlockTracer_LineOfSightTrace`.

**Pull request merged**